[SYCL] Do not internalize kernels when supporting dynamic linking

LU-JOHN opened the pull request:
Do not internalize kernels when supporting dynamic linking.  Kernels must be visible so that host code can find them.

sarnex commented on llvm/lib/SYCLLowerIR/ModuleSplitter.cpp:
```diff
@@ -667,11 +667,13 @@ void ModuleDesc::restoreLinkageOfDirectInvokeSimdTargets() {
     // visible outside of the current module.
     if (SupportDynamicLinking)
-      return canBeImportedFunction(*F);
+      return canBeImportedFunction(*F) ||
```
Can you explain how the kernel is being used in the host code in the failing case? It is just when we try to call it from the SYCL runtime, or some weird edge case?

LU-JOHN replied:
The free function kernels have names that start with double underscore, (e.g. `__sycl_kernel_ff_0`).  Thus, this function cannot be imported and in the previous code the symbol was internalized.   When this kernel was called by the host the symbol was not found and there was an error.

Note that in the user source free function kernels do not start with double underscores.  The implementation of free function kernels adds `__sycl_kernel_`.  The user source look like:

```
extern "C" SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(
    (ext::oneapi::experimental::single_task_kernel)) void ff_0(int *ptr,
                                                               int start,
                                                               int end) {
```

sarnex replied:
Ah okay, so this only happens with free function kernels, makes sense.

sarnex replied:
Should this check go inside `canBeImportedFunction` or should free function kernels not be imported but should not be internalized?

LU-JOHN replied:
My rationale for not including kernels in `canBeImportedFunction` is in the design doc sharedLibraries.md:

>   Note that kernel functions should not be included in `canBeImportedFunction` since kernels
>   are only callable by host code, and thus would never need to be imported into a device image.

sarnex replied:
ok, thanks

aelovikov-intel commented on sycl/test-e2e/KernelAndProgram/free_function_apis.cpp:
```diff
@@ -1,6 +1,8 @@
 // REQUIRES: aspect-usm_shared_allocations
 // RUN: %{build} -o %t.out
 // RUN: %{run} %t.out
+// RUN: %{build} -o %t.out -fsycl-allow-device-dependencies
+// RUN: %{run} %t.out
```
```suggestion
// RUN: %{build} -o %t2.out -fsycl-allow-device-dependencies
// RUN: %{run} %t2.out
```

same below.

LU-JOHN replied:
New temp names used for executables.

asudarsa commented on sycl/doc/design/SharedLibraries.md:
```diff
@@ -160,8 +160,9 @@ if the Function "can be imported".  A `canBeImportedFunction` is:
   2. Name does not start with "__"
-  3. Demangled name does not start with "__"
-  4. Must be a `SYCL_EXTERNAL` function
+  3. Is not a SPIRV, SYCL, or ESIMD builtin function
```
I suspect this document change is not related to the code change. right? Just wanted to make sure I understand this. Thanks

LU-JOHN replied:
The new point 3 is not related to the changes in this PR.   It is documenting a change that was previously done.

aelovikov-intel reviewed (approved):
How do you select which tests to modify?

LU-JOHN commented:
> How do you select which tests to modify?

I privately enabled -fsycl-allow-device-dependencies by default and made a list of all new failures.  These were the new failures fixed by this change.

aelovikov-intel commented:
> I privately enabled -fsycl-allow-device-dependencies by default and made a list of all new failures. These were the new failures fixed by this change.

Are we planning to flip the default in future? If so, we need a comment near these new RUN lines that they should be dropped once that happens. And if not, I'm not sure this is a good strategy for the tests for a feature...

LU-JOHN commented:
> > I privately enabled -fsycl-allow-device-dependencies by default and made a list of all new failures. These were the new failures fixed by this change.
> 
> Are we planning to flip the default in future? If so, we need a comment near these new RUN lines that they should be dropped once that happens. And if not, I'm not sure this is a good strategy for the tests for a feature...

Removed new RUN lines.  Added new testcase in DeviceDependencies directory to exercise this fix.

AlexeySachkov commented on llvm/lib/SYCLLowerIR/ModuleSplitter.cpp:
```diff
@@ -667,11 +667,13 @@ void ModuleDesc::restoreLinkageOfDirectInvokeSimdTargets() {
     if (SupportDynamicLinking)
-      return canBeImportedFunction(*F);
+      return canBeImportedFunction(*F) ||
+             F->getCallingConv() == CallingConv::SPIR_KERNEL;
```
I would swap the two checks, because calling convention check is cheaper and therefore we will be able to avoid calling more expensive `canBeImportedFunction` for some functions in a module

LU-JOHN replied:
Done.

AlexeySachkov commented on sycl/test-e2e/DeviceDependencies/free_function_kernels.cpp:
```diff
@@ -0,0 +1,270 @@
+  std::cout << "Test 0a: " << (PassA ? "PASS" : "FAIL") << std::endl;
+
+  bool PassB = false;
+#ifndef __SYCL_DEVICE_ONLY__
```
Unless that's a known issue with free function kernels that requires use to use `__SYCL_DEVICE_ONLY__` you should not be using such `#ifdef`s.

AlexeySachkov replied:
Fine by me to defer a resolution to a separate PR, but it would be good to leave some kind of comment/`TODO` maybe to note that

LU-JOHN replied:
Add TODO comment

AlexeySachkov commented on sycl/test-e2e/DeviceDependencies/free_function_kernels.cpp:
```diff
@@ -0,0 +1,270 @@
+  });
+  Queue.wait();
+  bool PassA = checkUSM(usmPtr, Range, Result);
+  std::cout << "Test 1a: " << (PassA ? "PASS" : "FAIL") << std::endl;
```
I don't think that our E2E tests should print anything in a positive scenario when they pass. The motivation for that is to reduce amount of I/O performed by tests and less data that has to be processed by `lit` framework. Logs for passed tests are hidden anyways, so they are of no use in our CI.

LU-JOHN replied:
Add TODO comment

AlexeySachkov replied:
I would rather immediately drop (almost) all `std::cout` lines, but I will leave that up to SYCL RT reviewers, who are code owners of the file

asudarsa reviewed (approved):
LGTM. Thanks

AlexeySachkov reviewed (approved):
`ModuleSplitter.cpp` and design doc changes LGTM
